Service Discovery External Services

longshorej opened the pull request:
This PR allows the service discovery component to accept external services declared via properties.

For instance,

`java ... -Drp.service-discovery.external-service-addresses.my-service-name-v1.0=http://127.0.0.1:7000`

fsat commented on service-discovery/src/main/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocator.scala:
```diff
@@ -18,44 +18,58 @@ package com.lightbend.rp.servicediscovery.scaladsl
+
+    settings.externalServiceAddresses.get(name) match {
+      case Some(services) =>
+        Future.successful(Some(services(ThreadLocalRandom.current.nextInt(services.length))))
```
I'd suggest allowing the caller to perform the service address selection themselves, i.e.

```
case class ServiceLocator {
  import ServiceLocator._
  
  def lookup(name: String, addressSelection: AddressSelection = AddressSelectionRandom): Future[Option[URI]] = 
    ServiceLocator.lookup(name, addressSelection)(as)
  // ...
}

object ServiceLocator {
  type AddressSelection = Seq[URI] => Option[URI]
  val AddressSelectionRandom: AddressSelection = (services => services(ThreadLocalRandom.current.nextInt(services.length))) 

  def lookup(name: String, addressSelection: AddressSelection = AddressSelectionRandom)(implicit as: ActorSystem): Future[Option[URI]] = {
    // ...
```

Then this line can be:

```
Future.successful(Some(addressSelection(services))))
```

longshorej replied:
Good idea - I did this for the Java API as well.

fsat commented on service-discovery/src/main/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocator.scala:
```diff
@@ -18,44 +18,58 @@ package com.lightbend.rp.servicediscovery.scaladsl
+                .mapTo[DnsServiceLocator.Addresses]
+            } yield
+              if (result.addresses.nonEmpty)
+                Some(addressToUri(result.addresses(ThreadLocalRandom.current.nextInt(result.addresses.size))))
```
Also this line can be:

```
Some(addressToUri(addressSelection(result.addresses))))
```

fsat commented on service-discovery/src/test/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocatorSpec.scala:
```diff
@@ -0,0 +1,65 @@
+        .lookup("has-two")
+        .map(e =>
+          e.contains(new URI("http://127.0.0.1:8000")) || e.contains(new URI("http://127.0.0.1:8001")) shouldBe true
+        )
```
Then your test can be:

```
ServiceLocator.lookup("has-two", addressSelection = (_.headOption)).futureValue.get shouldBe new URI("http://127.0.0.1:8000")
ServiceLocator.lookup("has-two", addressSelection = (_.lastOption)).futureValue.get shouldBe new URI("http://127.0.0.1:8001")
```

longshorej replied:
I just did this as two separate tests because I'm using `AsyncWordSpecLike` already.

fsat commented on service-discovery/src/test/scala/com/lightbend/rp/servicediscovery/scaladsl/ServiceLocatorSpec.scala:
```diff
@@ -0,0 +1,65 @@
+  with ImplicitSender
+  with AsyncWordSpecLike
+  with Matchers
+  with BeforeAndAfterAll {
```
Provided you have also extended:

```
with ScalaFutures
```

longshorej commented:
Thanks for the reviews. @fsat I've addressed your feedback (and also added similar API to the java API), and had @dbrinegar confirm so I'm going to merge in the interest of deadlines. Please feel free to open issues if I missed anything.
